File: go/feed-forward-iris-softmax-cross-entropy.go
package main

import ("encoding/csv"
        "fmt"
        "os"
        "gonum.org/v1/gonum/mat"
        "strconv"
        "math"
        "math/rand")

func removeDuplicates(elements []string) []string {
    encountered := map[string]bool{}
    result := []string{}
    for v := range elements {
        if encountered[elements[v]] == true {
        } else {
            encountered[elements[v]] = true
            result = append(result, elements[v])
        }
    }
    return result
}

// send address, printMatrix(&X)
func printMatrix(X *mat.Dense){
    fm := mat.Formatted(X, mat.Prefix("    "), mat.Squeeze())
    fmt.Printf("m = %4.2f", fm)
}

func indexOf(element string, elements []string) int{
    for i := 0; i < len(elements); i++ {
        if elements[i] == element {
            return i
        }
    }
    return -1
}

// generate ones, ones := generateConstant(150, 3, 1)
func generateConstant(x int, y int, constant float64) *mat.Dense{
    size := x * y
    matrix := make([]float64, size)
    for i := 0; i < size; i++ {
        matrix[i] = constant
    }
    return mat.NewDense(x, y, matrix)
}

func sigmoid(x float64, grad bool) float64{
    if grad{
        return sigmoid(x,false) * (1.0 - sigmoid(x,false))
    } else {
        return 1.0 / (1.0 + math.Exp(-x))
    }
}

func cross_entropy(x float64, y float64, grad bool) float64{
    if grad{
        return -(y / x) + (1 - y) / (1 - x)
    } else{
        return -y * math.Log(x) - (1 - y) * math.Log(1 - x)
    }
}

func calculateCrossEntropy(X *mat.Dense, Y *mat.Dense, grad bool) *mat.Dense{
    rows, columns := X.Dims()
    data := make([]float64, rows * columns)
    for i := 0; i < rows; i++ {
        for k := 0; k < columns; k++ {
            data[(i*columns)+k] = cross_entropy(X.At(i, k), Y.At(i, k), grad)
        }
    }
    return mat.NewDense(rows, columns, data)
}

func getMaxColumn(X *mat.Dense) *mat.Dense{
    rows, columns := X.Dims()
    data := make([]float64, rows)
    for i := 0; i < rows; i++ {
        data[i] = mat.Max(X.RowView(i))
    }
    m := generateConstant(rows, columns, 0)
    for i := 0; i < columns; i++ {
        m.SetCol(i, data)
    }
    return m
}

func getSumColumn(X *mat.Dense) *mat.Dense{
    rows, columns := X.Dims()
    data := make([]float64, rows)
    for i := 0; i < rows; i++ {
        data[i] = mat.Sum(X.RowView(i))
    }
    m := generateConstant(rows, columns, 0)
    for i := 0; i < columns; i++ {
        m.SetCol(i, data)
    }
    return m
}

func softmax(X *mat.Dense, grad bool) *mat.Dense{
    if grad{
        p := softmax(X, false)
        rows, columns := p.Dims()
        ones := generateConstant(rows, columns, 1)
        ones.Sub(ones, p)
        p.MulElem(p, ones)
        return p
    } else{
        max_columns := getMaxColumn(X)
        X.Sub(X, max_columns)
        applyExp := func(_, _ int, v float64) float64 { return math.Exp(v) }
        X.Apply(applyExp, X)
        sum_columns := getSumColumn(X)
        X.DivElem(X, sum_columns)
        return X
    }
}

func indexOfMax(arr []float64) int{
    max := arr[0];
    maxIndex := 0;
    for i := 1; i < len(arr); i++ {
        if (arr[i] > max) {
            maxIndex = i;
            max = arr[i];
        }
    }
    return maxIndex;
}

func getAccuracy(X *mat.Dense, Y *mat.Dense) float64{
    correct := 0
    rows, _ := X.Dims()
    for i := 0; i < rows; i++ {
        x := indexOfMax(X.RawRowView(i))
        y := indexOfMax(Y.RawRowView(i))
        if x == y{
            correct++
        }
    }
    return float64(correct) / float64(rows)
}

func randomMatrix(x int, y int) *mat.Dense{
    data := make([]float64, x * y)
    for i := range data {
        data[i] = rand.NormFloat64()
    }
    return mat.NewDense(x, y, data)
}

const EPOCH = 50
const LEARNING_RATE = 0.0001

func main() {
    file, _ := os.Open("iris.csv")
    defer file.Close()
    reader := csv.NewReader(file)
    record, _ := reader.ReadAll()
    record = record[1:]
    row_size := len(record)
    column_size := len(record[0][1:len(record[0])-1])
    iris := make([]float64, row_size * column_size)
    flowers := make([]string, row_size)
    
    for i := 0; i < row_size; i++ {
        for k := 0; k < column_size; k++ {
            val, _ := strconv.ParseFloat(record[i][k+1],64)
            iris[(i*column_size)+k] = val
        }
        flowers[i] = record[i][len(record[i])-1]
    }
    x_iris := mat.NewDense(row_size, column_size, iris)
    unique_flowers := removeDuplicates(flowers)
    
    // code our onehot-encoder
    onehot := mat.NewDense(row_size, len(unique_flowers), nil)
    for i := 0; i < row_size; i++ {
        onehot.Set(i, indexOf(flowers[i], unique_flowers), 1)
    }
    
    W1 := randomMatrix(column_size, 64)
    W2 := randomMatrix(64, 64)
    W3 := randomMatrix(64, len(unique_flowers))

    applySigmoid := func(_, _ int, v float64) float64 { return sigmoid(v,false) }
    applySigmoidDerivative := func(_, _ int, v float64) float64 { return sigmoid(v,true) }
    for iteration := 0; iteration < EPOCH; iteration++ {
        var a1 mat.Dense
        a1.Mul(x_iris, W1)
        z1 := new(mat.Dense)
        z1.Apply(applySigmoid, &a1)
        var a2 mat.Dense
        a2.Mul(z1, W2)
        z2 := new(mat.Dense)
        z2.Apply(applySigmoid, &a2)
        var a3 mat.Dense
        a3.Mul(z2, W3)
        y_hat := softmax(&a3,false)
        accuracy := getAccuracy(y_hat, onehot)
        err := calculateCrossEntropy(y_hat, onehot,false)
        cost := mat.Sum(err) / float64(column_size * row_size)
        fmt.Printf("epoch %d, cost %f, accuracy %f\n", iteration, cost, accuracy)
        dy_hat := calculateCrossEntropy(y_hat, onehot,true)
        da3 := softmax(&a3,true)
        da3.MulElem(da3, dy_hat)
        var dW3 mat.Dense
        dW3.Mul(z2.T(), da3)
        var dz2 mat.Dense
        dz2.Mul(da3, W3.T())
        da2 := new(mat.Dense)
        da2.Apply(applySigmoidDerivative, &a2)
        da2.MulElem(da2, &dz2)
        var dW2 mat.Dense
        dW2.Mul(z1.T(), da2)
        var dz1 mat.Dense
        dz1.Mul(da2, W2.T())
        da1 := new(mat.Dense)
        da1.Apply(applySigmoidDerivative, &a1)
        da1.MulElem(da1, &dz1)
        var dW1 mat.Dense
        dW1.Mul(x_iris.T(), da1)
        dW3.Scale(-LEARNING_RATE, &dW3)
        dW2.Scale(-LEARNING_RATE, &dW2)
        dW1.Scale(-LEARNING_RATE, &dW1)
        W3.Add(W3,&dW3)
        W2.Add(W2,&dW2)
        W1.Add(W1,&dW1)
    }
}

File: php/feed-forward-iris-softmax-cross-entropy.php
<?php
error_reporting(0);
require __DIR__ . '/vendor/autoload.php';

use MathPHP\LinearAlgebra\Vector;
use MathPHP\LinearAlgebra\Matrix;
use MathPHP\LinearAlgebra\MatrixFactory;

function generateZero($length){
    $array = array();
    for($i = 0; $i < $length; $i++) array_push($array, 0);
    return $array;
}

function randomMatrix($x, $y){
    $array = array();
    for($i = 0; $i < $x; $i++){
        $inner = array();
        for($k = 0; $k < $y; $k++){
            array_push($inner, mt_rand() / mt_getrandmax());
        }
        array_push($array,$inner);
    }
    return MatrixFactory::create($array);
}

function getMax($array){
    $max = $array[0];
    for($i = 1; $i < sizeof($array); $i++){
        if($array[$i] > $max) $max = $array[$i];
    }
    return $max;
}

function getSum($array){
    $sum = $array[0];
    for($i = 1; $i < sizeof($array); $i++) $sum += $array[$i];
    return $sum;
}

function getIndexMax($array){
    $max = 0;
    for($i = 1; $i < sizeof($array); $i++){
        if($array[$i] > $array[$max]) $max = $i;
    }
    return $max;
}

function getAccuracy($x, $y){
    $x = $x->getMatrix();
    $y = $y->getMatrix();
    $correct = 0;
    for($i = 0; $i < sizeof($x); $i++){
        $index_x = getIndexMax($x[$i]);
        $index_y = getIndexMax($y[$i]);
        if($index_x == $index_y) $correct++;
    }
    return $correct / sizeof($x);
}

function getMean($array){
    $sum = 0;
    for($i = 0; $i < $array->getM(); $i++) for($k = 0; $k < $array->getN(); $k++) $sum += $array[$i][$k];
    return $sum / ($array->getM() * $array->getN());
}

function getMaxColumn($m){
    $array = $m->getMatrix();
    $rows = $m->getM();
    $columns = $m->getN();
    $maxs = array();
    for($i = 0; $i < $rows; $i++) array_push($maxs, getMax($array[$i]));
    $array_max = array();
    for($i = 0; $i < $columns; $i++) array_push($array_max, $maxs);
    return MatrixFactory::create($array_max)->transpose();
}

function getSumColumn($m){
    $array = $m->getMatrix();
    $rows = $m->getM();
    $columns = $m->getN();
    $sums = array();
    for($i = 0; $i < $rows; $i++) array_push($sums, getSum($array[$i]));
    $array_sum = array();
    for($i = 0; $i < $columns; $i++) array_push($array_sum, $sums);
    return MatrixFactory::create($array_sum)->transpose();
}

function apply_function($m, $func){
    $m = $m->getMatrix();
    for($i = 0; $i < sizeof($m); $i++) for($k = 0; $k < sizeof($m[0]); $k++) $m[$i][$k] = $func($m[$i][$k]);
    return MatrixFactory::create($m);
}

function divide($a, $b){
    $a = $a->getMatrix();
    for($i = 0; $i < sizeof($a); $i++) for($k = 0; $k < sizeof($a[0]); $k++) $a[$i][$k] /= $b[$i][$k];
    return MatrixFactory::create($a);
}

function softmax($m, $grad){
    if($grad){
        $p = softmax($m,false);
        $ones = MatrixFactory::one($p->getM(), $p->getN());
        $ones = $ones->subtract($p);
        return $p->hadamardProduct($ones);
    }
    else{
        $max_columns = getMaxColumn($m);
        $m = $m->subtract($max_columns);
        $m = apply_function($m, exp);
        $sum_columns = getSumColumn($m);
        return divide($m, $sum_columns);
    }
}

function clipping($m, $min, $max){
    $m = $m->getMatrix();
    for($i = 0; $i < sizeof($m); $i++){
        for($k = 0; $k < sizeof($m[0]); $k++){
            if($m[$i][$k] < $min) $m[$i][$k] = $min;
            if($m[$i][$k] > $max) $m[$i][$k] = $max;
        }
    }
    return MatrixFactory::create($m);
}

function cross_entropy($x, $y, $grad){
    if($grad){
        $x = clipping($x, 1e-15, 1-1e-15);
        $x = $x->getMatrix();
        for($i = 0; $i < sizeof($x); $i++){
            for($k = 0; $k < sizeof($x[0]); $k++){
                $x[$i][$k] = -1 * ($y[$i][$k] / ($x[$i][$k]+1e-15)) + (1 - $y[$i][$k]) / (1 - $x[$i][$k]+1e-15);
            }
        }
        return MatrixFactory::create($x);
    }
    else{
        $x = clipping($x, 1e-15, 1-1e-15);
        $x = $x->getMatrix();
        for($i = 0; $i < sizeof($x); $i++){
            for($k = 0; $k < sizeof($x[0]); $k++){
                $x[$i][$k] = -1 * $y[$i][$k] * log($y[$i][$k]+1e-15) - (1 - $y[$i][$k]) * log(1 - $x[$i][$k]+1e-15);
            }
        }
        return MatrixFactory::create($x);
    }
}

function sigmoid($m, $grad){
    if($grad){
        $ones = MatrixFactory::one($m->getM(), $m->getN());
        $ones = $ones->subtract(sigmoid($m,false));
        return sigmoid($m,false)->hadamardProduct($ones);
    }
    else{
        $m = $m->getMatrix();
        for($i = 0; $i < sizeof($m); $i++) for($k = 0; $k < sizeof($m[0]); $k++) $m[$i][$k] = 1 / (1 + exp(-1 * $m[$i][$k]));
        return MatrixFactory::create($m);
    }
}

function derivative_cross_entropy_softmax($p, $y){
    $p = $p->getMatrix();
    for($i = 0; $i < sizeof($p); $i++){
        $p[$i][$y[$i]] -= 1;
    }
    return MatrixFactory::create($p);
}

$EPOCH = 100;
$LEARNING_RATE = 0.00005;
$iris = array();
$flowers = array();
$file = fopen("iris.csv","r");
$i = 0;

while(!feof($file)){
    $temp = fgetcsv($file);
    if($i > 0) {
        array_push($iris, array_slice($temp,1,-1));
        array_push($flowers, $temp[sizeof($temp)-1]);
    }
    $i++;
}
fclose($file);
$iris = array_slice($iris,0,-1);
$flowers = array_slice($flowers,0,-1);
$x_iris = MatrixFactory::create($iris);
$unique_flowers = array_values(array_unique($flowers));

$onehot = array();
$y_iris = array();
for($i = 0; $i < sizeof($iris); $i++) {
    $zeros = generateZero(sizeof($unique_flowers));
    $zeros[array_search($flowers[$i], $unique_flowers)] = 1;
    array_push($onehot, $zeros);
    array_push($y_iris, array_search($flowers[$i], $unique_flowers));
}
$onehot = MatrixFactory::create($onehot);
$W1 = randomMatrix($x_iris->getN(),64);
$W2 = randomMatrix(64,64);
$W3 = randomMatrix(64,sizeof($unique_flowers));

for($iteration = 0; $iteration < $EPOCH; $iteration++) {
    $a1 = $x_iris->multiply($W1);
    $z1 = sigmoid($a1,false);
    $a2 = $z1->multiply($W2);
    $z2 = sigmoid($a2,false);
    $a3 = $z2->multiply($W3);
    $y_hat = softmax($a3,false);
    $accuracy = getAccuracy($y_hat, $onehot);
    $cost = getMean(cross_entropy($y_hat,$onehot,false));
    $dy_hat = cross_entropy($y_hat,$onehot,true);
    $da3 = softmax($a3, true)->hadamardProduct($dy_hat);
    $dW3 = $z2->transpose()->multiply($da3);
    $dz2 = $da3->multiply($W3->transpose());
    $da2 = sigmoid($a2, true)->hadamardProduct($dz2);
    $dW2 = $z1->transpose()->multiply($da2);
    $dz1 = $da2->multiply($W2->transpose());
    $da1 = sigmoid($a1, true)->hadamardProduct($dz1);
    $dW1 = $x_iris->transpose()->multiply($da1);
    $W3 = $W3->add($dW3->scalarMultiply(-1*$LEARNING_RATE));
    $W2 = $W2->add($dW2->scalarMultiply(-1*$LEARNING_RATE));
    $W1 = $W1->add($dW1->scalarMultiply(-1*$LEARNING_RATE));
    printf("epoch %u, cost %f, accuracy %f\n",$iteration+1,$cost,$accuracy);
}
?>